File: 프로그래머스/3/43163. 단어 변환/단어 변환.js
function solution(begin, target, words){
    let head = 0;
    const q = [[begin, 0]];
    const visited = new Set([begin]);
    
    const isOneLetterDiff = (a, b) => {
        let cnt = 0;
        for(let i = 0 ; i < a.length ; i++){
            if(a[i] !== b[i]) cnt++;
        }
        return cnt === 1;
    }
    
    while(head < q.length){
        const [cur, step] = q[head++];
        
        if(cur === target) return step;
        
        for(const word of words){
            if(isOneLetterDiff(cur, word) && !visited.has(word)){
                visited.add(word);
                q.push([word, step + 1]);
            }
        }
    }
    
    return 0;
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.py
from collections import deque
def is_one_letter_diff(a, b):
    cnt = 0
    for x, y in zip(a, b):
        if x != y:
            cnt += 1
    return cnt == 1

def solution(begin, target, words):
    visited = set([begin])
    q = deque([[begin, 0]])
    
    while q:
        cur, step = q.popleft()
        if cur == target:
            return step
        for word in words:
            if is_one_letter_diff(word, cur) and word not in visited:
                visited.add(word)
                q.append([word, step+1])
    return 0